spark/spark-tensorflow-distributor/spark_tensorflow_distributor/mirrored_strategy_runner.py

```diff
@@ -0,0 +1,331 @@
+import logging
+import math
+import os
+import random
+import re
+import sys
+
+
+from pyspark.sql import SparkSession
+from pyspark.sql import SQLContext
+
+
+def _get_logger(name):
+    """
+    Gets a logger by name, or creates and configures it for the first time.
+    """
+    logger = logging.getLogger(name)
+    logger.setLevel(logging.INFO)
+    # If the logger is configured, skip the configure
+    if not logger.handlers and not logging.getLogger().handlers:
+        handler = logging.StreamHandler(sys.stderr)
+        logger.addHandler(handler)
+    return logger
+
+
+class MirroredStrategyRunner:
+    """
+    MirroredStrategyRunner runs TensorFlow deep learning training jobs on Spark clusters.
+    It trains synchronously by mirroring the model's variables among all the workers and
+    shares the gradient updates in a decentralized manner. MirroredStrategyRunner can take
+    a regular TensorFlow program with no special distributed training code and launch it
+    as a distributed training program.
+
+    .. note:: See more at https://www.tensorflow.org/guide/distributed_training
+    """
+    def __init__(self, num_slots, gpu_resource_name='gpu', use_custom_strategy=False):
+        """
+        :param num_slots: Total number of GPUs or CPU only Spark tasks that participate in distributed training.
+            When num_slots < 0, training is done in local mode on the Spark driver, and otherwise
+            training is distributed among the workers on the Spark cluster. For example,
+            num_slots = -4 means we train locally on 4 slots. If num_slots = 16 we train on the Spark cluster 
+            with 16 GPUs if doing GPU training, or with 16 Spark tasks if doing CPU training. num_slots cannot be 0.
+            Note that when doing CPU training, Spark will still be subject to any GPU-aware scheduling confs set in
+            the Spark configuration. Note also that for GPU training, num_slots will limit the number of GPUs used
+            for training even if more are available, so that exactly num_slots GPUs are used in total. Spark does not
+            restrict CPU cores for tasks and so for CPU training, num_slots rarely needs to be greater than the
+            number of workers and for local mode set num_slots=-1.
```

Comment:
Yes, since `num_slots` represent the number of spark task slots in CPU training, a user might want to say run two TensorFlow training workers and each Spark worker.